### PR TITLE
Add support for named unique, primary and foreign keys to SQLite3

### PR DESCRIPTION
**Pull request by carragom**
Attempt to solve #2051. Allows having named `unique`, `primary` and `foreign keys`. Both fluent and normal API should work except for `foreign keys` which have no fluent counterpart on any dialect AFAIK. This could break things for some user that might have been misusing the API. From my take on [this](https://github.com/tgriesser/knex/blob/0.13.0/test/unit/schema/sqlite3.js#L128-L135) test, in the current version of the SQLite dialect, the following code
```
table.primary('foo', 'bar')
```
Creates a primary key with two columns `foo` and `bar` when it should have created a primary key named `bar` with one column `foo` instead, as seen in other dialects like [MySQL](https://github.com/tgriesser/knex/blob/0.13.0/test/unit/schema/mysql.js#L176-L183).

Let's see if Travis likes it.

### CODE REVIEW

**Comment by elhigu** [comment]
Looks good to me.

**Comment by elhigu** [comment]
Thing that has worried me lately with many pull requests that tests are only covering generated queries, but they usually doesn't tests that those queries are actually doing the thing that they try to do correctly.

I would always like to have some integration test for this kind of changes, which actually tries the feature against real database. 

Now I'm worried that changed queries doesn't actually do what they are supposed to do and in worst case they break the code of many end users. 

I'll add requirement of integration tests against real DB to the contributing.md guide, since its better to have library working stable and well than having new features which are not tested well.

**Comment by carragom** [comment]
Well, I'm not sure an integration test would have prevented this specific problem. The integration test expected result would have been wrong just like the expected generated query is wrong but in the end, the integration test would have passed just like the unit test because the wrongly expected result is also valid. I think to prevent this from happening unit and integration test should be required, yes. But also these tests need to be validated extensively to make sure the expectation of the test is correct. Also, having a standard list of tests all with the same name would be really good. Currently, tests with the same intent have slightly different names on most dialects.

**Comment by elhigu** [comment]
> Well, I'm not sure an integration test would have prevented this specific problem. 

That comment of requirement of having integration test, had nothing to do with the test failure of this pull request :) Lack of them is just a thing that has been making feel uneasy already for a long time and that kind of tests makes sure that intended behaviour is not broken in the future by accident.

I would prefer leaving current integration test as is and just add a new simpler case which supports also sqlite. In that case also those alter column method tests will remain intact.

**Comment by carragom** [comment]
Ok, here we go with the integration tests. Let me know what you think.

**Comment by carragom** [comment]
Any idea about why is that test failing?. I see no relation between my changes and that specific test.

**Comment by elhigu** [comment]
@carragom it seems to be one annoying random failure that happens every now and then... I'll run the test again.

File: test/integration/schema/index.js [hunk 1]
```diff
@@ -576,5 +576,111 @@ module.exports = function(knex) {
...
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.primary('test', constraintName)
+            })
+          })
+          .then(function() {
+            return tr.schema.dropTableIfExists(tableName);
```
Comment (elhigu) [inline test/integration/schema/index.js]: Here you should try to drop constraint with raw query or do something else to make sure that given name was used for creating the named constraint.

File: test/integration/schema/index.js [hunk 2]
```diff
@@ -576,5 +576,111 @@ module.exports = function(knex) {
...
+              table.string('test');
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.primary(['test', 'test2'], constraintName)
+            })
+          });
+      });
```
Comment (elhigu) [inline test/integration/schema/index.js]: Here the same thing to check that constraint with name was created.

File: test/integration/schema/index.js [hunk 3]
```diff
@@ -576,5 +576,111 @@ module.exports = function(knex) {
...
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.unique('test', singleUniqueName);
+              table.unique(['test', 'test2'], multiUniqueName);
+            })
+          });
+      });
```
Comment (elhigu) [inline test/integration/schema/index.js]: The same kind of check needed

Comment (elhigu) [inline test/integration/schema/index.js]: maybe creating other constraint with the same name could also work for testing...

File: test/integration/schema/index.js [hunk 4]
```diff
@@ -576,5 +576,111 @@ module.exports = function(knex) {
...
+          .then(function() {
+            return tr.schema.createTable(joinTableName, function(table) {
+              table.uuid('user').references('id').inTable(userTableName)
+                .withKeyName(['fk', joinTableName, userTableName].join('-'));
+              table.uuid('group');
+              table.primary(['user', 'group']);
+              table.foreign('group', ['fk', joinTableName, groupTableName].join('-'))
+                .references('id').inTable(groupTableName);
+            });
+          });
```
Comment (elhigu) [inline test/integration/schema/index.js]: and here too.

**Comment by carragom** [comment]
Primary and unique tests should be good to go. Let me know what you think. I'll work on foreign keys sometime next week.

**Comment by carragom** [comment]
Forgot a console.log in there, I'll remove it when I fix the foreign key tests.

**Comment by carragom** [comment]
This PR is ready. I will rebase my branch to master but I'll like to wait for a response to my comment on #2087 first.

**Comment by elhigu** [comment]
@carragom I think, that comment has been answered. Are you still going to finish this?

**Comment by carragom** [comment]
@elhigu Your are right, this is ready to go, I should have time to finish this in the next couple of weeks.

**Comment by kibertoad** [comment]
@carragom Do you need any help with this?

**Comment by carragom** [comment]
I'm sorry for the delay on this. Short on time. @kibertoad if you can take over that would be great.

**Pull request closed without merging**